geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureSource.scala

```
@@ -0,0 +1,238 @@
+
+  /**
+   * The default behavior for getCount is to use estimated statistics.
+   * In most cases, this should be fairly close.
+   *
+   * Since users may want <b>exact</b> counts, there are two ways to force exact counts.
+   * First, one can set the System property "geomesa.force.count".
+   * Second, there is an EXACT_COUNT query hint.
+   *
+   * @param query query
+   * @return count
+   */
+  override def getCount(query: Query): Int = {
+    import org.locationtech.geomesa.index.conf.QueryHints.RichHints
+    import org.locationtech.geomesa.index.conf.QueryProperties.QUERY_EXACT_COUNT
+
+    val useExactCount = query.getHints.isExactCount.getOrElse(QUERY_EXACT_COUNT.get.toBoolean)
+    lazy val exactCount = ds.stats.getCount(getSchema, query.getFilter, exact = true).getOrElse(-1L)
+
+    val count = if (useExactCount) { exactCount } else {
+      ds.stats.getCount(getSchema, query.getFilter, exact = false).getOrElse(exactCount)
+    }
+    if (count > Int.MaxValue) Int.MaxValue else count.toInt
+  }
+
+  override def getBounds: ReferencedEnvelope = getBounds(new Query(sft.getTypeName, Filter.INCLUDE))
+
+  override def getBounds(query: Query): ReferencedEnvelope =
+    ds.stats.getBounds(getSchema, query.getFilter)
+
+  override def getQueryCapabilities = GeoMesaQueryCapabilities
+
+  override def getFeatures: SimpleFeatureCollection = getFeatures(Filter.INCLUDE)
+
+  override def getFeatures(filter: Filter): SimpleFeatureCollection =
+    getFeatures(new Query(sft.getTypeName, filter))
+
+  override def getFeatures(query: Query): SimpleFeatureCollection = collection(query, this)
+
+  override def getName: Name = getSchema.getName
+
+  override def getDataStore: DataStore = ds
+
+  override def getSupportedHints: java.util.Set[Key] = hints
+
+  override def getInfo: ResourceInfo = new DelegatingResourceInfo(this)
+
+  override def addFeatureListener(listener: FeatureListener): Unit = throw new NotImplementedError()
```

seems better to fail hard no?